File: game-server/src/main/java/work/fking/pangya/game/net/ClientProtocol.java
package work.fking.pangya.game.net;

public class ClientProtocol {

    private static final int PACKETS_SIZE = 512;
    private final ClientPacketType[] packets;
    private final ClientPacketHandler[] handlers;

    private ClientProtocol(ClientPacketType[] packets, ClientPacketHandler[] handlers) {
        this.packets = packets;
        this.handlers = handlers;
    }

    public static ClientProtocol create(ClientPacketType[] packets) {
        var packetMap = new ClientPacketType[PACKETS_SIZE];
        var handlers = new ClientPacketHandler[PACKETS_SIZE];

        for (var packet : packets) {
            packetMap[packet.id()] = packet;
            handlers[packet.id()] = packet.handler();
        }
        return new ClientProtocol(packetMap, handlers);
    }

    public ClientPacketHandler[] handlers() {
        return handlers;
    }

    public ClientPacketType forId(int packetId) {
        if (packetId < 0 || packetId >= PACKETS_SIZE) {
            return null;
        }
        return packets[packetId];
    }
}

File: game-server/src/main/kotlin/work/fking/pangya/game/net/ClientProtocol.kt
package work.fking.pangya.game.net

private const val PACKETS_SIZE = 512

class ClientProtocol(
    packetTypes: Array<ClientPacketType>
) {

    private val packets = arrayOfNulls<ClientPacketType>(PACKETS_SIZE)
    private val handlers = arrayOfNulls<ClientPacketHandler>(PACKETS_SIZE)

    init {
        for (packet in packetTypes) {
            packets[packet.id()] = packet
            handlers[packet.id()] = packet.handler()
        }
    }

    fun handlers(): Array<ClientPacketHandler?> {
        return handlers
    }

    fun forId(packetId: Int): ClientPacketType? {
        return if (packetId < 0 || packetId >= PACKETS_SIZE) {
            null
        } else packets[packetId]
    }

}
